aardvark/retrievers/runner.py

```diff
@@ -0,0 +1,169 @@
+        self.retrievers.append(r)
+
+    async def _run_retrievers(self, name: str):
+        """Run all registered retrievers."""
+        log.debug(f"creating {name}")
+        while True:
+            arn = await self.arn_queue.get()
+            log.debug(f"{name} retrieving data for {arn}")
+            data = {}
+            # Iterate through retrievers, passing the results from the previous to the next.
+            for r in self.retrievers:
+                data = await r.run(arn, data)
+            # TODO: store results
+            self.arn_queue.task_done()
+
+    async def _get_arns_for_account(self, account: str):
+        """Retrieve ARNs for roles, users, policies, and groups in an account and add them to the ARN queue."""
+        conn_details: Dict[str, str] = {
+            "account_number": account,
+            "assume_role": self.config["aws"]["rolename"].as_str(),
+            "session_name": "aardvark",
+            "region": self.config["aws"]["region"].as_str() or "us-east-1",
+            "arn_partition": self.config["aws"]["arn_partition"].as_str() or "aws",
+        }
+        client = await sync_to_async(boto3_cached_conn)("iam", service_type="client", **conn_details)
+
+        for role in list_roles(**conn_details):
+            await self.arn_queue.put(role["Arn"])
+
+        for user in list_users(**conn_details):
+            await self.arn_queue.put(user["Arn"])
+
+        for page in client.get_paginator("list_policies").paginate(Scope="Local"):
+            for policy in page["Policies"]:
+                await self.arn_queue.put(policy["Arn"])
+
+        for page in client.get_paginator("list_groups").paginate():
+            for group in page["Groups"]:
+                await self.arn_queue.put(group["Arn"])
+
+    async def _run_arn_lookup(self, name: str):
+        log.debug(f"creating {name}")
+        while True:
+            account = await self.account_queue.get()
+            log.debug(f"{name} retrieving ARNs for {account}")
+            await self._get_arns_for_account(account)
+            self.account_queue.task_done()
+
+    async def _get_swag_accounts(self) -> List[Dict]:
+        all_accounts: List[str] = []
+        try:
+            all_accounts: List[Dict] = self.swag.get_all(
+                self.swag_config["filter"].get()
+            )
+            swag_service = self.swag_config["service_enabled_requirement"].get()
+            if swag_service:
+                all_accounts = self.swag.get_service_enabled(
+                    swag_service, accounts_list=all_accounts
+                )
+        except (KeyError, InvalidSWAGDataException) as e:
+            log.error(
+                "Account names passed but SWAG not configured or unavailable: {}".format(
+                    e
+                )
+            )
+
+        return all_accounts
+
+    async def _queue_all_accounts(self):
+        for account in await self._get_swag_accounts():
```

Comment:
is this need to be await for _get_swag_accounts call? I don't see any asynchronous call in the function unless SWAG back-end does it.

Reply:
Wrapped the SWAG calls with `sync_to_async`